pkg/chunk/cache/redis_cache.go

```diff
@@ -0,0 +1,251 @@
+	f.DurationVar(&cfg.Expiration, prefix+"redis.expiration", 0, description+"How long keys stay in the redis.")
+	f.IntVar(&cfg.MaxIdleConns, prefix+"redis.max-idle-conns", 80, description+"Maximum number of idle connections in pool.")
+	f.IntVar(&cfg.MaxActiveConns, prefix+"redis.max-active-conns", 0, description+"Maximum number of active connections in pool.")
+}
+
+// NewRedisCache creates a new RedisCache
+func NewRedisCache(cfg RedisConfig, name string, client RedisClient) *RedisCache {
+	// client != nil in unit tests
+	if client == nil {
+		client = &redisClient{
+			pool: &redis.Pool{
+				MaxIdle:   cfg.MaxIdleConns,
+				MaxActive: cfg.MaxActiveConns,
+				Dial: func() (redis.Conn, error) {
+					c, err := redis.Dial("tcp", cfg.Endpoint)
+					if err != nil {
+						return nil, err
+					}
+					return c, err
+				},
+			},
+		}
+	}
+
+	cache := &RedisCache{
+		expiration: int(cfg.Expiration.Seconds()),
+		timeout:    cfg.Timeout,
+		name:       name,
+		client:     client,
+		requestDuration: observableVecCollector{
+			v: redisRequestDuration.MustCurryWith(prometheus.Labels{
+				"name": name,
+			}),
+		},
+	}
+
+	if err := cache.ping(context.Background()); err != nil {
+		level.Error(util.Logger).Log("msg", "error connecting to redis", "endpoint", cfg.Endpoint, "err", err)
+	}
+
+	return cache
+}
+
+// Fetch gets keys from the cache. The keys that are found must be in the order of the keys requested.
+func (c *RedisCache) Fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string) {
+	var data [][]byte
+	err := instr.CollectedRequest(ctx, "fetch", c.requestDuration, redisStatusCode, func(ctx context.Context) (err error) {
+		data, err = c.mget(ctx, keys)
+		return err
+	})
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "failed to get from redis", "name", c.name, "err", err)
+		missed = make([]string, len(keys))
+		copy(missed, keys)
+		return
+	}
+	for i, key := range keys {
+		if data[i] != nil {
+			found = append(found, key)
+			bufs = append(bufs, data[i])
+		} else {
+			missed = append(missed, key)
+		}
+	}
+	return
+}
+
+// Store stores the key in the cache.
+func (c *RedisCache) Store(ctx context.Context, keys []string, bufs [][]byte) {
+	err := instr.CollectedRequest(ctx, "store", c.requestDuration, redisStatusCode, func(ctx context.Context) error {
+		return c.mset(ctx, keys, bufs, c.expiration)
+	})
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "failed to put to redis", "name", c.name, "err", err)
+	}
+}
+
+// Stop stops the redis client.
+func (c *RedisCache) Stop() error {
+	return c.client.Close()
+}
+
+// mset adds key-value pairs to the cache.
+func (c *RedisCache) mset(ctx context.Context, keys []string, bufs [][]byte, ttl int) error {
+	res := make(chan error, 1)
+
+	conn := c.client.Connection()
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
```

Comment:
Why isn't this using the Redis library timeout facility? 
https://github.com/gomodule/redigo/blob/bb5990dc771ab24e169aaea90ee3b9ad50cffce9/redis/redis.go#L82